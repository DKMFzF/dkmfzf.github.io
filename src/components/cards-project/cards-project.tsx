import { FC } from 'react';
import { CardsProjectUI } from '@ui/index';
import { CardsProjectProps } from '../ui/cards-project/type';

export const CardsProject: FC<CardsProjectProps> = ({
	cards,
	transformStyle
}) => <CardsProjectUI cards={cards} transformStyle={transformStyle} />;
