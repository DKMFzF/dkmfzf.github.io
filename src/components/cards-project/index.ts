export { CardsProject } from './cards-project';
