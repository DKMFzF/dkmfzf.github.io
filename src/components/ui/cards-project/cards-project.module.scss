@use '~styles/variables' as *;

$premary-card-color: #cec2e6;

.cardsWrapper {
	position: absolute;
	font-family: $primary-font;
	will-change: transform;
	transform-origin: 0 0;
	user-select: none;
}

.card {
	position: absolute;
	z-index: 1;
	display: block;
	min-inline-size: 300px;
	overflow: hidden;
	color: $bg-primary-color;
	text-decoration: none;
	background: linear-gradient(135deg, #cec2e6 0%, #b3d5df 100%);
	border: 1px solid rgb(255 255 255 / 20%);
	border-radius: $section-border-radius;
	box-shadow: 0 4px 20px rgb(0 0 0 / 8%);
	cursor: pointer;

	&__titleContainer {
		padding-inline: 20px;
		padding-block: 20px;
		background-color: $bg-primary-color;
	}

	&__title {
		color: white;
		font-weight: 600;
		font-size: 36px;
		white-space: nowrap;
	}

	&__discriptionContainer {
		padding-inline: 20px;
		padding-block-end: 20px;
	}

	&__discription {
		display: block;
		margin-bottom: 15px;
		color: $bg-primary-color;
		font-size: 1rem;
		line-height: 1.5;
		opacity: $opacity;
	}

	&__btn {
		display: block;
		width: fit-content;
		padding: 10px;
		background-color: white;
		border-radius: 50%;
	}
}
