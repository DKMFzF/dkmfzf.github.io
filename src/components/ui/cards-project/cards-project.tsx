import React, { useRef } from 'react';
import gsap from 'gsap';
import { CardsProjectProps } from './type';
import styles from './cards-project.module.scss';

export const CardsProjectUI: React.FC<CardsProjectProps> = ({
	cards,
	transformStyle
}) => {
	const cardRefs = useRef<Map<number, HTMLAnchorElement>>(new Map());

	return (
		<div
			className={styles.cardsWrapper}
			style={{ transform: transformStyle }}
		>
			{cards.map((card) => (
				<a
					target='_blank'
					key={card.id}
					ref={(el) => el && cardRefs.current.set(card.id, el)}
					href={card.link}
					className={styles.card}
					style={{
						left: `${card.x}px`,
						top: `${card.y}px`,
						transform: `rotate(${card.rotation}deg)`
					}}
					onMouseEnter={() => {
						const el = cardRefs.current.get(card.id);
						if (el) {
							gsap.to(el, {
								y: -15,
								scale: 1.07,
								boxShadow: '0 15px 25px rgba(0,0,0,0.2)',
								duration: 0.3,
								ease: 'power3.out'
							});
							el.style.zIndex = '2';
						}
					}}
					onMouseLeave={() => {
						const el = cardRefs.current.get(card.id);
						if (el) {
							gsap.to(el, {
								y: 0,
								scale: 1,
								boxShadow: '0 4px 10px rgba(0,0,0,0.1)',
								duration: 0.3,
								ease: 'power3.out'
							});
							el.style.zIndex = '1';
						}
					}}
					onMouseDown={(e) => e.stopPropagation()}
				>
					<div className={styles.card__titleContainer}>
						<h3 className={styles.card__title}>{card.title}</h3>
					</div>
					<div className={styles.card__discriptionContainer}>
						<p className={styles.card__discription}>
							{card.content}
						</p>
						<span className={styles.card__btn}>-{'>'}</span>
					</div>
				</a>
			))}
		</div>
	);
};
