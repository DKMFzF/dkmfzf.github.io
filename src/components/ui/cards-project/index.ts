export { CardsProjectUI } from './cards-project';
