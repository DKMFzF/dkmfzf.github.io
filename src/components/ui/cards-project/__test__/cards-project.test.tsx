import { testRenderComponent } from '@utils-testing';
import { CardsProjectUI } from '../cards-project';

describe('[CardsProjectUI]', () => {
	it('Renders card-project correctly and matches snapshot', () =>
		testRenderComponent(
			<CardsProjectUI
				cards={[
					{
						id: 1,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 110,
						y: 110,
						rotation: 10
					},
					{
						id: 2,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 220,
						y: 220,
						rotation: 20
					},
					{
						id: 3,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 330,
						y: 330,
						rotation: -10
					},
					{
						id: 4,
						title: 'string',
						content: 'string',
						link: 'string',
						x: 440,
						y: 440,
						rotation: -20
					}
				]}
				transformStyle='test-styles'
			/>
		));
});
