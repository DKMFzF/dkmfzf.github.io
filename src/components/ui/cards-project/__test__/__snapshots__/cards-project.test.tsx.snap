// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[CardsProjectUI] Renders card-project correctly and matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="cards-wrapper"
    style="transform: test-styles;"
  >
    <li
      class="cards__element"
    >
      <a
        class="card"
        href="string"
        style="left: 110px; top: 110px; transform: rotate(10deg);"
        target="_blank"
      >
        <div
          class="card__title-container"
        >
          <h3
            class="card__title"
          >
            string
          </h3>
        </div>
        <div
          class="card__discription-container"
        >
          <p
            class="card__discription"
          >
            string
          </p>
          <span
            class="card__btn"
          >
            &gt;
          </span>
        </div>
      </a>
    </li>
    <li
      class="cards__element"
    >
      <a
        class="card"
        href="string"
        style="left: 220px; top: 220px; transform: rotate(20deg);"
        target="_blank"
      >
        <div
          class="card__title-container"
        >
          <h3
            class="card__title"
          >
            string
          </h3>
        </div>
        <div
          class="card__discription-container"
        >
          <p
            class="card__discription"
          >
            string
          </p>
          <span
            class="card__btn"
          >
            &gt;
          </span>
        </div>
      </a>
    </li>
    <li
      class="cards__element"
    >
      <a
        class="card"
        href="string"
        style="left: 330px; top: 330px; transform: rotate(-10deg);"
        target="_blank"
      >
        <div
          class="card__title-container"
        >
          <h3
            class="card__title"
          >
            string
          </h3>
        </div>
        <div
          class="card__discription-container"
        >
          <p
            class="card__discription"
          >
            string
          </p>
          <span
            class="card__btn"
          >
            &gt;
          </span>
        </div>
      </a>
    </li>
    <li
      class="cards__element"
    >
      <a
        class="card"
        href="string"
        style="left: 440px; top: 440px; transform: rotate(-20deg);"
        target="_blank"
      >
        <div
          class="card__title-container"
        >
          <h3
            class="card__title"
          >
            string
          </h3>
        </div>
        <div
          class="card__discription-container"
        >
          <p
            class="card__discription"
          >
            string
          </p>
          <span
            class="card__btn"
          >
            &gt;
          </span>
        </div>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
